messaging/WFLY-10518_artemis_no_local_broker.adoc

```diff
@@ -0,0 +1,166 @@
+= [WFLY-10518] Connections to remote Artemis broker must not require local Artemis instance
+:author:            Jeff Mesnil
+:email:             jmesnil@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          messaging,jms,openshift
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Connections to remote Artemis-based broker (including A-MQ 7) are handled in
+WildFly by the
+https://wildscribe.github.io/WildFly/13.0/subsystem/messaging-activemq/server/pooled-connection-factory/index.html[pooled-connection-factory]
+resource that is in the `messaging-activemq`.
+
+This `pooled-connection-factory` resource can be used to connect to the local
+ Artemis server that "owns" it or to a remote Artemis broker.
+
+However, this `pooled-connection-factory` resource always require the presence
+ of a local Artemis server even when it connects only to a remote Artemis broker.
+This use case will become more prevalent as WildFly is deployed on OpenShift to
+connect to a remote A-MQ 7 broker.
+
+The presence of a local Artemis server has impact on the memory and CPU
+ footprint of WildFly image on OpenShift.
+This RFE will improve WildFly so that a connection to a remote Artemis-based
+ broker will no longer require the presence of a local instance of Artemis.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-10518[WFLY-10518]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1014[EAP7-1014]
+* https://issues.jboss.org/browse/EAP7-1006[EAP7-1006] is a RFE related to Java EE7 JMS resource
+definitions that will benefit from this RFE.
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:mnovak@redhat.com[Miroslav Novak]
+
+=== Affected Projects or Components
+
+* This is a internal restructuring of the `messaging-activemq` subsystem that requires no changes
+from Artemis or other messaging components.
```

Comment:
It will result in changes to our OpenShift images though.

Reply:
@jmesnil @bstansberry will be there some migration path for
a) current OpenShift deployments
b) classical .zip - e.g. some equivalent for migrate operation
?

Reply:
For bare metal or any cloud situation where the user is bringing their own standalone.xml, I don't see a need for a migrate operation. We've used migrate operations when we have removed something and the user must make a config change to have a working server. That's not the case here. The old way still works fine.

The migration tool has gone beyond that a bit in trying to adapt the user's old config to whatever "best practices" are in the new release, but I doubt that's feasible here. I don't know how the tool would know that the embedded broker in the old config isn't needed.

For these kinds of situations it sounds to me like a docs issue. Perhaps  I'll feel differently when we get a bunch of migration code contributors who are up for long term maintenance too. ;)

As for OpenShift deployments moving from EAP 7.1 to 7.2 where they are relying on our tooling to generate their config document we'll need to work out a migration path so people who need the embedded broker don't lose the ability to have it. All these things are controlled by env vars, so perhaps something like this:

USE_OLD_WAY -- default is true in 7.2, no value results in a WARN to stdout saying default will change to false in 7.3 and recommending setting....
USE_NEW_WAY -- default is false in 7.2, true in 7.3

Of course those aren't the real env var names.

It's possible in our 7.3 images we'd not generate an embedded broker config at all any more. Our basic direction (and not just us engineers) is that an embedded broker is a poor fit on the cloud. The reasons for that go well beyond resource utilization. Our docs already state that users shouldn't use an embedded broker.

@luck3y fyi.

Reply:
@tremes this feature will affect OpenShift

Reply:
@tremes Note though that we (aka @luck3y) are likely not going to develop any image scripting changes to configure standalone-openshift.xml this way before the WF 14 feature freeze. If you want to experiment with this in an image before then let's chat.